MapboxVisionTests/VisionManagerTests.swift

```diff
@@ -0,0 +1,139 @@
+@testable import MapboxVision
+import XCTest
+
+class VisionManagerTests: XCTestCase {
+    var visionManager: VisionManager?
+    var dependencies: VisionDependencies?
+
+    override func setUp() {
+        super.setUp()
+
+        let native = MockNative()
+
+        dependencies = VisionDependencies(
+            native: native,
+            synchronizer: MockSynchronizable(),
+            recorder: MockSessionRecorder(),
+            dataProvider: MockDataProvider(),
+            deviceInfo: DeviceInfoProvider()
+        )
+
+        self.visionManager = VisionManager(dependencies: dependencies!, videoSource: MockVideoSource())
+    }
+
+    func test_AssertOnVisionManagerDeallocWithoutDestroy() {
+        // Description:
+        // We called destroy in destructors of VisionManager, BaseVisionManager, VisionManagerNative and VisionManagerBaseNative,
+        // so VisionManagerBaseNative's destroy method called multiple times, so we tried to destroy already destroyed resources,
+        // which caused an EXC_BAD_ACCESS.
+        //
+        // Nevertheless the logic was sort of right, it should do assertFailure on deinit without destroy was called
+        //
+        // Given
+        // object of VisionManager
+        // When
+        // we deallocate it without destroy
+        // Then
+        // Should be assertion failure
+        MVTAssertAssertionFailure(
+            self.visionManager = nil,
+            "VisionManager should call assertioFailure when we release the object without calling destroy"
+        )
+    }
+
+    func test_NoAssertOnVisionManagerDeallocAfterDestroy() {
+        // Given
+        // object of VisionManager
+
+        // When
+        // we call destroy and release the instance
+        self.visionManager?.destroy()
+
+        // Then
+        // Shouldn't be assertion failure
+        MVTAssertNoAssertionFailure(
+            self.visionManager = nil,
+            "VisionManager should be successfully deallocated after calling destroy and releasing the object"
+        )
+    }
+
+    func test_RecordingPermissions() {
+        // Given
+        // VisionManager with default country
+
+        guard let visionManager = self.visionManager, let recorder = dependencies?.recorder as? MockSessionRecorder else {
+            XCTAssert(false)
+            return
+        }
+
+        // When
+        // We set countries
+
+        visionManager.onCountryUpdated(.USA)
+        visionManager.onCountryUpdated(.UK)
+        visionManager.onCountryUpdated(.china)
+        visionManager.onCountryUpdated(.other)
+        visionManager.onCountryUpdated(.unknown)
+
+        // Then
+        // we expect that visionManager will call SessionRecorder's method `startInternal` for all countries but China and that there
+        // will no be any other calls to SessionRecorder. For China we expect `stop` method to be called
```

Comment:
nit:

As far as I understand from comment we're trying to test 2 cases here:
- we have expected actions for each case;
- amount of final actions exactly equal to amount of expected actions);

If so probably we can add additional XCTAssert below. One for the first case and another one for the second.

Reply:
I think it's redundant, @chezzdev WDYT?

Reply:
I guess we're testing "we have expected actions for each case" here and the second point follows from the first one. Comparing arrays tests both cases BTW.